src/Passes/Source/RuleTransformationPass/RulePass.cpp

```diff
@@ -0,0 +1,863 @@
+            }
+        }
+
+        for (auto& function : module)
+        {
+            bool is_active = false;
+
+            // Checking if the function is referenced by a global variable
+            for (auto user : function.users())
+            {
+                // If the user is active, then it should be expected that the function is also active
+                if (isActive(user))
+                {
+                    is_active = true;
+                    break;
+                }
+            }
+
+            if (is_active || function.hasFnAttribute(config_.entryPointAttr()))
+            {
+                // Marking function as active
+                active_pieces_.insert(&function);
+
+                // Detecting active code
+                runOnFunction(function, [this](llvm::Value* value, DeletableInstructions& modifier) {
+                    return detectActiveCode(value, modifier);
+                });
+            }
+        }
+
+        applyReplacements();
+    }
+
+    void RuleTransformationPass::applyReplacements()
+    {
+        // Applying all replacements
+
+        std::unordered_set<llvm::Value*> already_removed;
+        for (auto it = replacements_.rbegin(); it != replacements_.rend(); ++it)
+        {
+            auto instr1 = llvm::dyn_cast<llvm::Instruction>(it->first);
+
+            // Checking if by accident the same instruction was added
+            if (already_removed.find(instr1) != already_removed.end())
+            {
+                throw std::runtime_error("Instruction was already removed.");
+            }
+            already_removed.insert(instr1);
+
+            if (instr1 == nullptr)
+            {
+                llvm::outs() << "; WARNING: cannot deal with non-instruction replacements\n";
+                continue;
+            }
+
+            // Cheking if have a replacement for the instruction
+            if (it->second != nullptr)
+            {
+                // ... if so, we just replace it,
+                auto instr2 = llvm::dyn_cast<llvm::Instruction>(it->second);
+                if (instr2 == nullptr)
+                {
+                    llvm::outs() << "; WARNING: cannot replace instruction with non-instruction\n";
+                    continue;
+                }
+                llvm::ReplaceInstWithInst(instr1, instr2);
+            }
+            else
+            {
+                // ... otherwise we delete the the instruction
+                // Removing all uses
+                if (!instr1->use_empty())
+                {
+                    auto type = instr1->getType();
+                    if (type != nullptr)
+                    {
+                        instr1->replaceAllUsesWith(llvm::UndefValue::get(type));
+                    }
+                }
+
+                // And finally we delete the instruction
+                if (instr1->use_empty())
+                {
+                    instr1->eraseFromParent();
+                }
```

Comment:
Can you add a few words of a comment about in which case the `instr1->use_empty()` can return `false` in this `if`?
(I would expect it always to return `true`)

Reply:
Up to you. (To me `assert()` sounds more appropriate based on what I understand from your words)